README.md

```diff
@@ -2,302 +2,708 @@
+# optional custom error message
+field(:salutation).validate(:format, /^Mr\/s\./, "must start with Mr/s.")
 ```
 
-### :options array
+### :email
 
-Declare allowed values in an array. Values not in the options will be ignored or use default value.
+```ruby
+field(:business_email).validate(:email)
+```
+
+### :required
+
+Check that the key exists in the input.
+
+```ruby
+field(:name).required
+
+# same as
+field(:name).validate(:required)
+```
+
+Note that _required_ does not validate that the value is not empty. Use _present_ for that.
+
+### :present
+
+Check that the key exists and the value is not blank.
 
 ```ruby
-class OrdersSearch
-  include Parametric::Params
-  param :sort, 'Sort', options: ['updated_on-desc', 'updated_on-asc'], default: 'updated_on-desc'
+field(:name).present
+
+# same as
+field(:name).validate(:present)
+```
+
+If the value is a `String`, it validates that it's not blank. If an `Array`, it checks that it's not empty. Otherwise it checks that the value is not `nil`.
+
+### :declared
+
+Check that a key exists in the input, or stop any further validations otherwise.
+This is useful when chained to other validations. For example:
+
+```ruby
+field(:name).declared.present
+```
+
+The example above will check that the value is not empty, but only if the key exists. If the key doesn't exist no validations will run.
+
+### :gt
+
+Validate that the value is greater than a number
+
+```ruby
+field(:age).validate(:gt, 21)
+```
+
+### :lt
+
+Validate that the value is less than a number
+
+```ruby
+field(:age).validate(:lt, 21)
+```
+
+### :options
+
+Pass allowed values for a field
+
+```ruby
+field(:status).options(["draft", "published"])
+
+# Same as
+field(:status).validate(:options, ["draft", "published"])
+```
+
+### :split
+
+Split comma-separated string values into an array.
+Useful for parsing comma-separated query-string parameters.
+
+```ruby
+field(:status).policy(:split) # turns "pending,confirmed" into ["pending", "confirmed"]
+```
+
+## Custom policies
+
+You can also register your own custom policy objects. A policy must implement the following methods:
+
+```ruby
+class MyPolicy
+  # Validation error message, if invalid
+  def message
+    'is invalid'
+  end
+
+  # Whether or not to validate and coerce this value
+  # if false, no other policies will be run on the field
+  def eligible?(value, key, payload)
+    true
+  end
+
+  # Transform the value
+  def coerce(value, key, context)
+    value
+  end
+
+  # Is the value valid?
+  def valid?(value, key, payload)
+    true
+  end
 end
 ```
 
-### :multiple values
+You can register your policy with:
 
-`:multiple` values are separated on "," and treated as arrays.
+```ruby
+Parametric.policy :my_policy, MyPolicy
+```
+
+And then refer to it by name when declaring your schema fields
 
 ```ruby
-class OrdersSearch
-  include Parametric::Params
-  param :status, 'Order status', multiple: true
+field(:title).policy(:my_policy)
```

Comment:
Custom types can actually be invoked with `#type(:my_type)`, since `#type` is just an alias of `#policy`, but I totally agree with this and was thinking about it. My only concern was that by allowing users to basically add any public methods they want (via method_missing) it could end up making it harder to me to extend the "official" public API in the future because of name clashes. But I like how succinct it makes it look so I'll give it a go.